### PR TITLE
Pass over the accessibility tree section

### PR DESCRIPTION
**Pull request by pavpanchekha**
This PR moves the accessibility tree section to go before we start implementing the screen reader in earnest, and makes some minor code cleanups.

I still need to pass over the "screen-reader" section, which was broken by some of these changes.

### CODE REVIEW

File: book/accessibility.md [hunk 1]
```diff
@@ -1405,26 +1405,180 @@ colors is something that page authors also have to pay attention to.
...
+also sometimes useful for situations where the user shouldn't be
+looking at the screen (such as driving), or for devices with no
+screen.] screen-reader software is typically used instead. The name
+kind of explains it all: screen-reader software reads the text on the
+screen, so that users know what it says without having to see it.
+
+So: what should we say to the user? There are basically two big
+challenges we must overcome.
+
+First, web pages contain visual hints besides text, that we need to
```
Comment (chrishtr) [inline book/accessibility.md:1420]:
```suggestion
First, web pages contain visual hints besides text that we need to
```

File: book/accessibility.md [hunk 2]
```diff
@@ -1405,26 +1405,180 @@ colors is something that page authors also have to pay attention to.
...
+And second, when listening to a screen-reader, the user must be able
+to direct the browser to the part of the page that interests
+them.[^fast] For example, the user might want to skip headers and
+navigation menus, or even skip most of the page until they get to a
+paragraph of interest. But once they've reached the part of the page
+of interest to them, they may want it read to them, and if some
+sentence or phrase is particularly complex, they may want the
+screen-reader to re-read it.
+
+You see an example[^imagine] of screen-reader navigation in this talk,
```
Comment (chrishtr) [inline book/accessibility.md:1434]:
```suggestion
You can see an example[^imagine] of screen-reader navigation in this talk,
```

File: book/accessibility.md [hunk 3]
```diff
@@ -1405,26 +1405,180 @@ colors is something that page authors also have to pay attention to.
...
+specifically the segment from 2:36--3:54:[^whole-talk]
+
+<iframe width="560" height="315" src="https://www.youtube-nocookie.com/embed/qi0tY60Hd6M?start=159" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
+
+[^whole-talk]: The whole talk is recommended; it has great examples of
+    using accessibility technology.
+
+[^fast]: Though though many people who rely on screen-readers learn to
```
Comment (chrishtr) [inline book/accessibility.md:1442]:
```suggestion
[^fast]: Though many people who rely on screen-readers learn to
```

File: book/accessibility.md [hunk 4]
```diff
@@ -1405,26 +1405,180 @@ colors is something that page authors also have to pay attention to.
...
+    dense medium than vision.
+
+[^imagine] I encourage you to test out your operating system's
+    built-in screen reader to get a feel for what screen-reader
+    navigation is like. On macOS, type Cmd-Fn-F5 to turn on Voice
+    Over; on Windows, type Win-Ctrl-Enter or Win-Enter to start
+    Narrator. Both are largely used via keyboard shortcuts that you
+    can look up.
+    
+To support all this, screen readers structure the page as a tree. The
```
Comment (chrishtr) [inline book/accessibility.md:1453]:
```suggestion
To support all this, browsers structure the page as a tree and use it to respond to screen reader APIs (these APIs are generally usually implemented outside the browser). The
```

File: book/accessibility.md [hunk 5]
```diff
@@ -1405,26 +1405,180 @@ colors is something that page authors also have to pay attention to.
...
+screen-reader users. The browser therefore builds a separate
+[accessibility tree][at] to support screen-reader navigation.
+
+[at]: https://developer.mozilla.org/en-US/docs/Glossary/Accessibility_tree
+
+[^invisible-example]: For example, using `opacity:0`. There are
+several other ways in real browsers that elements can be made
+invisible, such as with the `visibility` or `display` CSS properties.
+
+The accessibility tree is built in a rendering phase just after layout:
```
Comment (chrishtr) [inline book/accessibility.md:1473]:
```suggestion
Let's implement the accessibility tree. It is built in a rendering phase just after layout:
```

File: book/accessibility.md [hunk 6]
```diff
@@ -1405,26 +1405,180 @@ colors is something that page authors also have to pay attention to.
...
+        if child.role != "none":
+            self.children.append(child)
+            parent = child
+        else:
+            parent = self
+        for child_node in node.children:
+            parent.build_internal(child_node)
+```
+
+The user can now direct the screen-reader to walks up or down this
```
Comment (chrishtr) [inline book/accessibility.md:1568]:
```suggestion
The user can now direct the screen-reader to walk up or down this
```

File: book/accessibility.md [hunk 7]
```diff
@@ -1405,26 +1405,180 @@ colors is something that page authors also have to pay attention to.
...
+            elif node.tag == "html":
+                self.role = "document"
+            elif is_focusable(node):
+                self.role = "focusable"
+            else:
+                self.role = "none"
+```
+
+To build the accessibility tree, we just recursively walk the HTML
+tree, adding all nodes whose role isn't `none`:
```
Comment (chrishtr) [inline book/accessibility.md:1549]:
```suggestion
tree, adding all nodes whose role isn't `none` (note that a regular `<div>` is therefore skipped, for example):
```

File: book/accessibility.md [hunk 8]
```diff
@@ -1478,6 +1619,107 @@ def speak_text(text):
...
+def speak_text(text):
+    print("SPEAK:", text)
+```
+
+Then you can see what the browser *would have* spoken out loud in the
+standard output.
+
+:::
+
+To start with, we'll want a keybinding that turns the screen-reader on
```
Comment (chrishtr) [inline book/accessibility.md:1641]:
```suggestion
To start with, we'll want a key binding that turns the screen-reader on
```

File: book/accessibility.md [hunk 9]
```diff
@@ -1478,6 +1619,107 @@ def speak_text(text):
...
+        self.lock.acquire(blocking=True)
+        self.accessibility_is_on = not self.accessibility_is_on
+        active_tab = self.tabs[self.active_tab]
+        task = Task(active_tab.toggle_accessibility)
+        active_tab.task_runner.schedule_task(task)
+        self.lock.release()
+```
+
+The `Tab`, in turn, executes the `speak_task` method if accessibility
+is on, which is actually produce sound:
```
Comment (chrishtr) [inline book/accessibility.md:1674]:
```suggestion
is on, which is what actually produces sound:
```

**Comment by pavpanchekha** [comment]
I came up with a (hacky) fix for the tests, so I'm going to mark this mergable and put the screen section in its own PR.

File: book/accessibility.md [hunk 10]
```diff
@@ -1405,26 +1405,182 @@ colors is something that page authors also have to pay attention to.
...
+    medium than vision.
+
+[^imagine] I encourage you to test out your operating system's
+    built-in screen reader to get a feel for what screen-reader
+    navigation is like. On macOS, type Cmd-Fn-F5 to turn on Voice
+    Over; on Windows, type Win-Ctrl-Enter or Win-Enter to start
+    Narrator. Both are largely used via keyboard shortcuts that you
+    can look up.
+    
+To support all this, browers structure the page as a tree and use that
```
Comment (chrishtr) [inline book/accessibility.md:1453]:
```suggestion
To support all this, browsers structure the page as a tree and use that
```

**Review by chrishtr — APPROVED** [review]

**Pull request merged**